src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/OnceInitializedOnceDisposedUnderLockAsync.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.ProjectSystem
+{
+    /// <summary>
+    ///     Provides an implementation of <see cref="OnceInitializedOnceDisposedAsync"/> that lets 
+    ///     implementors protect themselves from being disposed while doing work.
+    /// </summary>
+    /// <remarks>
+    ///     <see cref="OnceInitializedOnceDisposed"/> lets implementors prevent themselves from being disposed
+    ///     by locking <see cref="OnceInitializedOnceDisposed.SyncObject"/>. This class provides a similar 
+    ///     mechanism by passing a delegate into <see cref="ExecuteUnderLockAsync"/>.
+    /// </remarks>
+    internal abstract class OnceInitializedOnceDisposedUnderLockAsync : OnceInitializedOnceDisposedAsync
```

Comment:
>OnceInitializedOnceDisposedUnderLockAsync [](start = 28, length = 41)

This might want to go to CPS for everyone to reuse?

Reply:
I'm raising it with them - I want to make sure it "works" correctly and consume it before moving it over to them.